vault/logical_system_test.go

```diff
@@ -6450,3 +6451,808 @@ func TestSystemBackend_pluginRuntime_CannotDeleteRuntimeWithReferencingPlugins(t
+			assert.Contains(t, resp.Data, "key_info", testcase.name)
+			assert.IsType(t, map[string]any{}, resp.Data["key_info"], testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any), "000", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "title", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "type", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "authenticated", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "start_time", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "end_time", testcase.name)
+			assert.Contains(t, resp.Data["key_info"].(map[string]any)["000"], "active", testcase.name)
+		} else {
+			assert.NotContains(t, resp.Data, "keys", testcase.name)
+		}
+
+		if testcase.expectErrorInData {
+			assert.Contains(t, resp.Data, "error", testcase.name)
+		} else {
+			assert.NotContains(t, resp.Data, "error", testcase.name)
+		}
+	}
+
+	// Finally, test when the underlying storage returns an error
+	uiConfig.barrierStorage = &testingStorage{
+		listFails: true,
+	}
+
+	resp, err := backend.handleListCustomMessages(context.Background(), &logical.Request{}, &framework.FieldData{
+		Schema: fieldSchemas,
+		Raw:    map[string]any{},
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, resp)
+	assert.NotNil(t, resp.Data)
+	assert.Contains(t, resp.Data, "error")
+}
+
+// TestHandleCreateCustomMessage verifies the proper functioning of the
+// (*SystemBackend).handleCreateCustomMessage method. The test focuses on
+// missing and invalid request parameter (framework.FieldData) and errors in the
+// underlying logical.Storage.
+func TestHandleCreateCustomMessage(t *testing.T) {
+	// Setup a system backend
+	uiConfig := &UIConfig{
+		barrierStorage: &logical.InmemStorage{},
+	}
+
+	backend := &SystemBackend{
+		Core: &Core{
+			uiConfig: uiConfig,
+		},
+	}
+
+	fieldSchemas := map[string]*framework.FieldSchema{
+		"title": {
+			Type: framework.TypeString,
+		},
+		"message": {
+			Type: framework.TypeString,
+		},
+		"authenticated": {
+			Type: framework.TypeBool,
+		},
+		"type": {
+			Type: framework.TypeString,
+		},
+		"start_time": {
+			Type: framework.TypeTime,
+		},
+		"end_time": {
+			Type: framework.TypeTime,
+		},
+		"link": {
+			Type: framework.TypeMap,
+		},
+		"options": {
+			Type: framework.TypeMap,
+		},
+	}
+
+	// the standard map of request parameters containing all required parameters
+	// with valid values. The test cases will make a copy of this map and modify
+	// it to test different conditions.
+	fieldRaw := map[string]any{
+		"title":         "title",
+		"message":       "message",
+		"authenticated": "true",
+		"type":          "banner",
+		"start_time":    "2023-01-01T00:00:00Z",
+		"end_time":      "2100-01-01T00:00:00Z",
+	}
+
+	testcases := []struct {
```

Comment:
I think some test cases would be good for end time before start time, and end time equal to start time, etc

Reply:
This is a good point 💯 , I was thinking that the validation would occur in the UI but this is foolish since the API could be used directly. I've pushed another commit that adds a validation function that's invoked by both the `(*UIConfig).CreateCustomMessage` and `(*UIConfig).UpdateCustomMessage` methods and I've also adjusted the tests to add a unit test for the validation function and add test cases in the `TestCreateCustomMessage` and `TestUpdateCustomMessage` testing functions.

I don't think we need to cover this case in this testing function since the purpose here is to cover the `(*SystemBackend).handleCreateCustomMessage` method which already has a testcase to cover any error being returned by `(*UIConfig).CreateCustomMessage`; whether the maximum number of custom messages exist, the times are reversed, or some other error occurred in the underlying storage layer, it's all seen the same way by handleCreateCustomMessage.

Reply:
I think the additional coverage can't hurt, but yeah, it does make sense to focus testing the validation on the lower level function. Thanks for adding this!